source-code/core2/src/plugin/api.ts

```diff
@@ -0,0 +1,86 @@
+import type { InlangApp } from "../app/api.js"
+import type { LanguageTag } from "../languageTag.js"
+import type { LintRule } from "../lint/api.js"
+import type { Message } from "../messages/schema.js"
+
+/**
+ * Proposal 1
+ *
+ * Setup function everything in one.
+ *
+ * + simple to use
+ * - no access to static props.
+ *   if a plugin crashes during setup -> complicated to show plugin with "id" crashed.
+ */
+export type Plugin_Proposal_1<Options extends Record<string, string> = Record<string, never>> =
+	(args: { options: Options; inlang: InlangApp }) => {
+		id: `${string}.${string}`
+		displayName: string
+		extendLanguageTags?: () => LanguageTag[]
+		loadMessages?: () => Message[]
+		saveMessages?: (args: { messages: Message[] }) => void
+		addLintRules?: () => LintRule[]
+	}
+
+export const example1: Plugin_Proposal_1<{ pathPattern: string }> = ({ options, inlang }) => {
+	return {
+		id: "inlang.myPlugin",
+		displayName: "My Plugin",
+		extendLanguageTags: () => {
+			return ["en-US"]
+		},
+		loadMessages: () => {
+			for (const languageTag of inlang.config.languageTags) {
+				console.log(languageTag + options.pathPattern)
+			}
+			return []
+		},
+		saveMessages: ({ messages }) => {
+			console.log(messages)
+		},
+	}
+}
+
+/**
+ * Proposal 2
+ *
+ * Split static meta data from setup function.
+ *
+ * + access to static without crash chance
+ * - arguably not so nice API.
+ */
+export type Plugin_Proposal_2<Options extends Record<string, string> = Record<string, never>> = {
+	meta: {
+		id: `${string}.${string}`
+		displayName: string
+	}
+	setup: (args: { options: Options; inlang: InlangApp }) => {
+		extendLanguageTags?: () => LanguageTag[]
+		loadMessages?: () => Message[]
```

Comment:
I would split lint rules and messages plugins. This API will grow once we have video audio etc. APIs, making it less clear "what do I need to provide, when I want to write a lint rule"

Reply:
Distinguishing different sort of plugins seems wrong. 

Let plugins use the plugin API however they want and need. If you only want to add a lint rule, only use `addLintRule`. The concept of having one "entrypoint" to extend inlang, namely creating a plugin, is superior to "you can create a lint rule via a lint rule plugin", "you can create a message plugin via a message plugin", etc. 

No, just "you can create a plugin". "In a plugin you can add lint rules, or define how messages should be loaded and saved". etc.